packages/flutter/lib/src/widgets/overlay.dart

```diff
@@ -959,3 +923,1325 @@ class _RenderTheatre extends RenderBox with ContainerRenderObjectMixin<RenderBox
+    super._activate(child);
+    _overlayChildRenderBox = child;
+  }
+
+  @override
+  void _deactivate(_RenderDeferredLayoutBox child) {
+    assert(_debugNotDisposed());
+    super._deactivate(child);
+    _overlayChildRenderBox = null;
+  }
+
+  @override
+  String toString() => '$_showTimestamp';
+}
+
+class _RenderTheatreMarker extends InheritedWidget {
+  const _RenderTheatreMarker({
+    required this.theatre,
+    required this.overlayEntryWidgetState,
+    required super.child,
+  });
+
+  final _RenderTheatre theatre;
+  final _OverlayEntryWidgetState overlayEntryWidgetState;
+
+  @override
+  bool updateShouldNotify(_RenderTheatreMarker oldWidget) {
+    return oldWidget.theatre != theatre
+        || oldWidget.overlayEntryWidgetState != overlayEntryWidgetState;
+  }
+
+  @override
+  InheritedElement createElement() => _RenderTheatreMarkerElement(this);
+
+  static _RenderTheatreMarker? of(BuildContext context, { required bool targetRootOverlay }) {
+    final _RenderTheatreMarkerElement? element = context.getElementForInheritedWidgetOfExactType<_RenderTheatreMarker>() as _RenderTheatreMarkerElement?;
+    if (element == null) {
+      return null;
+    }
+    final _RenderTheatreMarkerElement dependency = targetRootOverlay ? element._topRenderTheatre : element;
+    return context.dependOnInheritedElement(dependency) as _RenderTheatreMarker;
+  }
+}
+
+class _RenderTheatreMarkerElement extends InheritedElement {
+  _RenderTheatreMarkerElement(super.widget);
+
+  _RenderTheatreMarkerElement get _topRenderTheatre {
+    _RenderTheatreMarkerElement? ancestor;
+    visitAncestorElements((Element element) {
+      assert(ancestor == null);
+      ancestor = element.getElementForInheritedWidgetOfExactType<_RenderTheatreMarker>() as _RenderTheatreMarkerElement?;
+      return false;
+    });
+    return ancestor?._topRenderTheatre ?? this;
+  }
+}
+
+class _OverlayPortalState extends State<OverlayPortal> {
+  // The developer must call `show` to reveal the overlay so we can get the
+  // timestamp of the user interaction for sorting.
+  _TopmostLocation? _location;
+  int? _showTimestamp;
+  // See https://github.com/dart-lang/sdk/issues/41717.
+  static int wallTime = kIsWeb
+    ? -9007199254740992 // -2^53
+    : -1 << 63;
+
+  @override
+  void initState() {
+    super.initState();
+    _setupController(widget.controller);
+  }
+
+  void _setupController(OverlayPortalController controller) {
+    assert(
+      controller._attachTarget == null || controller._attachTarget == this,
+      'Failed to attach $controller to $this. It is already attached to ${controller._attachTarget}'
+    );
+    final int? controllerTimeStamp = controller._showTimestamp;
+    final int? timeStamp = _showTimestamp;
+    if (timeStamp == null || (controllerTimeStamp != null && controllerTimeStamp > timeStamp)) {
+      _showTimestamp = controllerTimeStamp;
+    }
+    controller._showTimestamp = null;
```

Comment:
in general i would recommend using a term like `zIndex` or `order` rather than `timestamp` since the latter usually means a `DateTime`. Someone coming to this code would be very confused to find that none of these time stamps have any relationship to the wall clock.

Reply:
should be less confusing now that the method that does the increment is called `_now()`?

Reply:
This API usually does not lead to a one-frame delay in the Overlay UI. `controller.{show(),hide()}` are meant to be called before the persistent frame callback phase so it's safe to `setState`. If overlayChild's visibility depends on an ancestor widget (e.g. an `InheritedWidget`) then people are supposed to use the (yet to be exposed) declarative API.

Reply:
I didn't mean the UI, I meant the SetState take one frame to update the location in the _OverlayPortal which I assume is why there is a timestamp so that you can tell the order of the setstate for different OverlayPortals. What I suggest is to hook the show/hide to the Element of _OverlayPortal so that it can store itself in a cache some where in overlay element so that you don't need the time stamp to tell the insertion order